@here/olp-sdk-core/lib/utils/MultiPartUploadWrapper.ts

```diff
@@ -177,9 +221,211 @@ export class MultiPartUploadWrapper {
+            blobVersion: opts.blobVersion,
+            catalogHrn: opts.catalogHrn,
+            abotrSignal
+        });
+
+        const startMultipartResponse = await request.startMultipartUpload({
+            contentType: opts.contentType,
+            handle: opts.handle,
+            layerId: opts.layerId,
+            billingTag: opts.billingTag,
+            contentEncoding: opts.contentEncoding
+        });
+
+        if (!this.isStartMultipartResponseValid(startMultipartResponse)) {
+            throw new Error(
+                "Failed to start multipart upload process, bad response."
+            );
+        }
+
+        opts.onStart &&
+            opts.onStart({
+                dataSize,
+                multipartToken: startMultipartResponse.multipartToken,
+                multipartUrl: startMultipartResponse.uploadPartUrl,
+                multipartStatusUrl: startMultipartResponse.statusUrl
+            });
+
+        const chunkSize = this.validateChunkSize(opts.chunkSizeMB);
+        const totalChunks = Math.ceil(dataSize / chunkSize);
+
+        let unreadBytes = dataSize;
+        let offset = 0;
+        let chunkNumber = 0;
+        let uploadedChunks = 0;
+
+        while (unreadBytes > 0) {
+            const promises = [];
+
+            while (promises.length < parallelRequests) {
+                const buffer = await blobData.readBytes(offset, chunkSize);
+
+                if (buffer.byteLength === 0) {
+                    break;
+                }
+
+                chunkNumber++;
+                offset += buffer.byteLength;
+                unreadBytes -= buffer.byteLength;
+
+                promises.push(
+                    request
+                        .uploadPart({
+                            data: buffer,
+                            layerId: opts.layerId,
+                            multipartToken:
+                                startMultipartResponse.multipartToken,
+                            billingTag: opts.billingTag,
+                            contentLength: buffer.byteLength,
+                            contentType: opts.contentType,
+                            partNumber: chunkNumber,
+                            url: startMultipartResponse.uploadPartUrl
+                        })
+                        .then(uploadPartResponse => {
+                            parts.push({
+                                id: uploadPartResponse.id,
+                                number: chunkNumber
+                            });
+
+                            uploadedChunks++;
+
+                            opts.onStatus &&
+                                opts.onStatus({
+                                    chunkId: uploadPartResponse.id,
+                                    chunkNumber,
+                                    chunkSize: buffer.byteLength,
+                                    totalChunks,
+                                    uploadedChunks
```

Comment:
Also, i would probably make a copy of this after incrementing to be sure you do not give the user a false value.

Reply:
Pls, see my comment above.